src/stream/src/common/table/state_table.rs

```diff
@@ -754,9 +754,17 @@ where
         }
     }
 
-    pub fn update_watermark(&mut self, watermark: ScalarImpl) {
+    /// Update watermark for state cleaning.
+    ///
+    /// # Arguments
+    ///
+    /// * `watermark` - Latest watermark received.
+    /// * `eager_cleaning` - Whether to clean up the state table eagerly.
+    pub fn update_watermark(&mut self, watermark: ScalarImpl, eager_cleaning: bool) {
         trace!(table_id = %self.table_id, watermark = ?watermark, "update watermark");
-        self.cur_watermark = Some(watermark);
+        if eager_cleaning || self.watermark_buffer_strategy.apply() {
```

Comment:
I think `eager_cleaning` could be absorbed by `watermark_buffer_strategy`

Reply:
Maybe we can leave this for now, and see if it's necessary in the future. (Actually I `eager_cleaning` may just be a temporary workaround for the lack of `table.range_delete()`)

Reply:
> Should we still call `apply` (or even `reset`) on the strategy if `eager_cleaning`?

Good point. I think for now `self.watermark_buffer_strategy.apply() || eager_cleaning` may be enough.